cirq-core/cirq/transformers/target_gatesets/compilation_target_gateset.py

```diff
@@ -29,15 +29,49 @@
     import cirq
 
 
-def _create_transformer_with_kwargs(func: 'cirq.TRANSFORMER', **kwargs) -> 'cirq.TRANSFORMER':
-    """Hack to capture additional keyword arguments to transformers while preserving mypy type."""
+def create_transformer_with_kwargs(transformer: 'cirq.TRANSFORMER', **kwargs) -> 'cirq.TRANSFORMER':
+    """Method to capture additional keyword arguments to transformers while preserving mypy type.
+
+    Returns a `cirq.TRANSFORMER` which, when called with a circuit and transformer context, is
+    equivalent to calling `transformer(circuit, context=context, **kwargs)`. It is often useful to
+    capture keyword arguments of a transformer before passing them as an argument to an API that
+    expects `cirq.TRANSFORMER`. For example:
+
+    >>> def run_transformers(transformers: List[cirq.TRANSFORMER]):
+    >>>     for transformer in transformers:
+    >>>         transformer(circuit, context=context)
+    >>>
+    >>> transformers: List[cirq.TRANSFORMER] = []
+    >>> transformers.append(
+    >>>     cirq.create_transformer_with_kwargs(
+    >>>         cirq.expand_composite, no_decomp=lambda op: cirq.num_qubits(op) <= 2
+    >>>     )
+    >>> )
+    >>> transformers.append(cirq.create_transformer_with_kwargs(cirq.merge_k_qubit_unitaries, k=2))
+    >>> run_transformers(transformers)
+
+
+    Args:
+         transformer: A `cirq.TRANSFORMER` for which additional kwargs should be captured.
+         **kwargs: The keyword arguments which should be captured and passed to `transformer`.
+
+    Returns:
+        A `cirq.TRANSFORMER` method `transformer_with_kwargs`, s.t. executing
+        `transformer_with_kwargs(circuit, context=context)` is equivalent to executing
+        `transformer(circuit, context=context, **kwargs)`.
+
+    Raises:
+        SyntaxError: if **kwargs contain a 'context'.
+    """
+    if 'context' in kwargs:
+        raise SyntaxError('**kwargs to be captured must not contain `context`.')
 
-    def transformer(
+    def transformer_with_kwargs(
         circuit: 'cirq.AbstractCircuit', *, context: Optional['cirq.TransformerContext'] = None
     ) -> 'cirq.AbstractCircuit':
-        return func(circuit, context=context, **kwargs)  # type: ignore
+        return transformer(circuit, context=context, **kwargs)  # type: ignore
```

Comment:
Why do we need the type ignore? I guess it's because `cirq.TRANSFORMER` is a [callable protocol](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols) whose `__call__` signature does not allow for extra kwargs. Might we be able to remove the type ignore by extending the signature with explicit kwargs?

In any case, there should be a comment explaining the reason for the ignore and linking to a mypy issue if there is one open.

Reply:
> I guess it's because cirq.TRANSFORMER is a [callable protocol](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols) whose __call__ signature does not allow for extra kwargs.

Yes, that's right. 

> Might we be able to remove the type ignore by extending the signature with explicit kwargs?

This is not possible because function arguments are contravariant, so adding `**kwargs` to the signature of the protocol would mean that every transformer would need to support a `**kwargs : Any`; which is bad. 

> In any case, there should be a comment explaining the reason for the ignore and linking to a mypy issue if there is one open.

Added a comment.